articles/best-practices/rules.md

```diff
@@ -1,129 +1,571 @@
+:::
+
+Asynchronous execution will result in a (JavaScript) callback being executed after the asynchronous operation is complete. This callback is typically fired at some point after the main (synchronous) body of a JavaScript function completes. If a rule is making use of asynchronous processing then a call to the (Auth0) supplied [`callback`](#callback-function) function must be deferred to the point where asynchronous processing completes - and must be the final thing called. The (Auth0) supplied `callback` function must be called only once; calling the function more than once within a rule will lead to unpredictable results and/or errors.
+
+### `context` object
+
+The [`context`](/rules/references/context-object) object provides information about the context in which a rule is run (such as client identifier, connection name, session identifier, request context, protocol, etc). Using the context object, a rule can determine the reason for execution. For example, as illustrated in the sample fragment below, [`context.clientID`](/rules/references/context-object#properties-of-the-context-object) as well as [`context.protocol`](/rules/references/context-object#properties-of-the-context-object) can be used to implement conditional processing to determine when rule logic is executed. The sample also shows some best practices for [exception handling](#exceptions), use of [`npm` modules](#npm-modules) (for `Promise` style processing), and the [`callback`](#callback-object) object. 
+
+```js
+  switch (context.protocol) {
+    case 'redirect-callback':
+      callback(null, user, context);
+    	break;
+
+    default: {
+      user.app_metadata = user.app_metadata || {};
+      switch(context.clientID) {
+        case configuration.PROFILE_CLIENT: {
+          user.user_metadata = user.user_metadata || {};
+          Promise.resolve(new 
+            Promise(function (resolve, reject) {
+              switch (context.request.query.audience) {
+                case configuration.PROFILE_AUDIENCE: {
+                  switch (context.connection) {
+		  			  .
+					  .
+                  }
+                } break;
+			  .
+			  .
+            })
+          )
+          .then(function () {
+		      .
+		      .
+          })
+          .catch(function (error) {
+            callback(new UnauthorizedError(“unauthorized”), user, context);
+          });
+        } break;
+          
+        default:
+          callback(null, user, context);
+          break;
+      
+    } break; 
+```
+
+::: warning
+It’s recommended best practice to [avoid using contextual logic for Multi-Factor Authentication(MFA)](?). For example, **serious security flaws** can surface if use of MFA is predicated on `context.request.query.prompt === 'none'`. Additionally, the contents of the `context` object is **security sensitive**, so you should [**not** directly pass the object to any external or 3rd party service](don-t-send-entire-context-object-to-external-services).
```

Comment:
Missing link: `[avoid using contextual logic for Multi-Factor Authentication(MFA)](?)`

Reply:
Updated links and content syntax